site/content/api/Credentials-API.md

```diff
@@ -0,0 +1,316 @@
++++
+title = "Credentials API"
+weight = 435
++++
+
+The *Credentials API* is used by *Protocol Adapters* to retrieve credentials used to authenticate *Devices* connecting to the adapter. In particular, the API supports the storage, look up and deletion of *shared secrets* which are often used by IoT devices by means of *username/password* based authentication schemes.
+<!--more-->
+
+Credentials are of a certain *type* which indicates which authentication mechanism the credentials can be used with. Each set of credentials also contains an *authentication identity* which is the identity claimed by the device during authentication. This authentication identity is usually different from the *logical* ID the device has been registered with using Hono's [Registration API]({{< ref "api/Device-Registration-API.md" >}}). Multiple sets of credentials (including arbitrary *authentication identities*) can be registered for each *logical* device ID.
+
+Note, however, that in real world applications the device credentials will probably be managed by a dedicated identity management system. In such a case an implementation of Hono's Credentials API can be created which maps the AMQP 1.0 based API calls to the identity management system's (remote) API.
+
+The Credentials API is defined by means of AMQP 1.0 message exchanges, i.e. a client needs to connect to Hono using an AMQP 1.0 client in order to invoke operations of the API as described in the following sections.
+
+# Preconditions
+
+The preconditions for invoking any of the Credential API's operations are as follows:
+
+1. Client has established an AMQP connection to Hono.
+2. Client has established an AMQP link in role *sender* with Hono using target address `credentials/${tenant_id}`. This link is used by the client to send commands to Hono.
+3. Client has established an AMQP link in role *receiver* with Hono using source address `credentials/${tenant_id}/${reply-to}` where *reply-to* may be any arbitrary string chosen by the client. This link is used by the client to receive responses to the requests it has sent to Hono. This link's source address is also referred to as the *reply-to* address for the request messages.
```

Comment:
This is how we did it with the Registration API. IIRC we ran into problems with JMS clients ...